### PR TITLE
REGR: to_stata tried to remove file before closing it

### PR DESCRIPTION
**Pull request by twoertwein**
Spin-off from #39047:

If an error occurred during `to_stata`, the code tried to first remove the file and then close the file handle:

1. removing the file failed on Windows (it is still opened)
2. It created an empty file when closing the handle

### CODE REVIEW

**Comment by jreback** [comment]
this is a regression (i think so?). put a note on 1.2.1

**Comment by twoertwein** [comment]
yes, you are right! I extended a test to make sure that the file doesn't exists and added a whatsnew entry.

**Review by jreback — APPROVED** [review]
excellent ping on greenish

**Comment by twoertwein** [comment]
The failing test will not fail in 'real life': the test uses `tm.ensure_clean()` to ensure that the file will not exist. Unfortunately, this means that we have an open file handle that `to_stata` cannot close. I will look into re-writing `ensure_clean()`.

File: doc/source/whatsnew/v1.2.1.rst [hunk 1]
```diff
@@ -32,6 +32,7 @@ Fixed regressions
 - Fixed regression in :meth:`DataFrame.groupby` when aggregating an :class:`ExtensionDType` that could fail for non-numeric values (:issue:`38980`)
 - Fixed regression in :meth:`DataFrame.loc.__setitem__` raising ``KeyError`` with :class:`MultiIndex` and list-like columns indexer enlarging :class:`DataFrame` (:issue:`39147`)
 - Fixed regression in comparisons between ``NaT`` and ``datetime.date`` objects incorrectly returning ``True`` (:issue:`39151`)
+- :meth:`to_state` did not remove written file when an error occured (:issue:`39202`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.1.rst:35]: sorry can you fix typo : to_stata also this should be DataFrame.to_stata

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
also some failures

**Comment by twoertwein** [comment]
I think `to_stata` is the only `to_*` method that tries to remove created files in case of an error. Is that something other methods should also try to do?

I'm not sure but I could imagine that `__exit__` in `IOHandles` might be able to determine whether an exception from within pandas's io part was raised: if yes, try to delete the files. This would probably only cover local files (no fsspec strings) and only work for `to_*` methods that use `with get_handles(...)` (some methods have to use it without the context manager and explicitly call `close` - `IOHandles` wouldn't know whether an exception occurred.

**Comment by jreback** [comment]
we should remove it if it's corrupt - however this might be hard to tell

so ok leaving it

**Comment by twoertwein** [comment]
@jreback green'ish

The new `ensure_clean` has the drawback (I see it as an advantage) that the code using it needs to make sure that no file handle references the temporary file. If an open file handle exists, the test on windows will fail. That's why I had to fix a few tests in this PR.

File: doc/source/whatsnew/v1.2.1.rst [hunk 2]
```diff
@@ -33,6 +33,7 @@ Fixed regressions
 - Fixed regression in :meth:`DataFrame.groupby` when aggregating an :class:`ExtensionDType` that could fail for non-numeric values (:issue:`38980`)
 - Fixed regression in :meth:`DataFrame.loc.__setitem__` raising ``KeyError`` with :class:`MultiIndex` and list-like columns indexer enlarging :class:`DataFrame` (:issue:`39147`)
 - Fixed regression in comparisons between ``NaT`` and ``datetime.date`` objects incorrectly returning ``True`` (:issue:`39151`)
+- :meth:`DataFrame.to_state` did not remove created file when an error occured (:issue:`39202`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.2.1.rst:36]: Nit: `DataFrame.to_state` -> `DataFrame.to_stata`

File: pandas/_testing/contexts.py [hunk 3]
```diff
@@ -73,66 +77,48 @@ def setTZ(tz):
 
 
 @contextmanager
-def ensure_clean(filename=None, return_filelike=False, **kwargs):
+def ensure_clean(filename="", return_filelike: bool = False, **kwargs: Any):
```
Comment (jreback) [inline pandas/_testing/contexts.py:80]: should revert to `None` right?

**Review by jreback — COMMENTED** [review]
lgtm comment, ping on green

**Comment by twoertwein** [comment]
@jreback  green'ish, failed test seem to be unrelated

**Comment by simonjayhawkins** [comment]
> failed test seem to be unrelated

yep, known flaky test and pre-commit issue

**Comment by simonjayhawkins** [comment]
one question before this is merged. How public-ish is tm.ensure_clean?

**Comment by jreback** [comment]
> one question before this is merged. How public-ish is tm.ensure_clean?

its not

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @twoertwein

**Comment by jreback** [comment]
@meeseeksdev backport 1.2.x

**Comment by jreback** [comment]
@twoertwein would you mind putting up the backport PR

**Pull request merged**